java/ql/src/semmle/code/java/dataflow/NullGuards.qll

```diff
@@ -104,10 +121,13 @@ predicate clearlyNotNull(SsaVariable v, Expr reason) {
     clearlyNotNull(captured, reason)
   )
   or
-  exists(Field f |
-    v.getSourceVariable().getVariable() = f and
-    f.isFinal() and
-    f.getInitializer() = clearlyNotNullExpr(reason)
+  exists(Variable var |
+    v.getSourceVariable().getVariable() = var and
+    var.isFinal() and
+    var.getAnAssignedValue() = clearlyNotNullExpr(reason) and
+    forall(Expr assignedValue | assignedValue = var.getAnAssignedValue() |
+      assignedValue = clearlyNotNullExpr()
+    )
```

Comment:
I believe this could be a `forex` statement: 

```suggestion
    forex(Expr assignedValue | assignedValue = var.getAnAssignedValue() |
      assignedValue = clearlyNotNullExpr()
    )
```

Reply:
I think this entire disjunct rewrite looks a bit curious. Why the change from `Field` to `Variable`? We ought to be able to handle all non-fields more precisely through SSA. What's the idea behind this change?

Reply:
> I believe this could be a `forex` statement:

I used `forall` here to report the `reason`; with your proposal `reason` would not be bound. Note that for `SwitchExpr` I tried binding the reason in the `forall` but that did not work, therefore I wrote it as separate formula here as well.

---

> I think this entire disjunct rewrite looks a bit curious

Do you mean usage of `getAnAssignedValue()`? The reason is that `getInitializer()` is too restrictive and does not cover a lot of cases, e.g. consider:
```java
class Test
  final String s;

  public Test(boolean b) {
    if (b) {
      s = "a";
    } else {
      s = "b";
    }
  }
}
```
Here `s` cannot be `null` because all assigned values are non-`null`, even though they do not appear as part of the initializer. But that does not matter since you cannot access an uninitialized `final` variable, see [JLS 16 §16](https://docs.oracle.com/javase/specs/jls/se16/html/jls-16.html).

---

> We ought to be able to handle all non-fields more precisely through SSA.

I think here it does not matter due to not being able to access an unassigned `final` variable, see above. However, at least at the current state reducing this back to `Field` fails finding `n3` of the included test.
I am not familiar enough with SSA, so if you think it should be solved using that, feel free to improve the pull request in that regard.

Reply:
> Do you mean usage of `getAnAssignedValue()`

Yes. It would be cleaner to add a case for phi-nodes. But this predicate isn't intended to cover everything that we think isn't null - it's intended to cover everything that _clearly_ isn't null.
Are these changes motivated by concrete query FPs in real code?

Reply:
> But this predicate isn't intended to cover everything that we think isn't null - it's intended to cover everything that _clearly_ isn't null.

When all `getAnAssignedValue()` expressions are clearly not null, then any read access of the variable has a clearly non-null result, see JLS 16 §16 linked above.
Checking only `getInitializer()` misses a lot cases where a variable or field is not directly assigned a value in the initializer, see Java source code example above.

(Note that this ignores reflection here since the previous code did not consider it either.)